dvc/scm.py

```diff
@@ -132,54 +132,73 @@ def resolve_rev(scm: "Git", rev: str) -> str:
         raise RevError(str(exc))
 
 
+def last_n_commits(
```

Comment:
Add this new function for the use of `list/push/pull/remove`. As we only have two sets of experiments retrieve flags (`rev`+`num`/`all-commits`), and the `all-commits` are well supported by our current utils. We only need this to get our experiment's collections logic unified instead of using the heavy `iter_revs` below.

Reply:
@pmrowla What do you think of it?
1. Now in  `exp list/push/pull/remove` we only support `--rev`, `-n` and `--all-commits`. For `all-commits` the process of it is quite different from any other args (we don't need to collect conditions), iter over all exps are just enough for it. A good example is in `remove` where the `--all-commits` is simpler than `--rev`, I don't think it is a good design to unify them ( first collect all revs, then match them for every exps).
2. Currently, `iter_revs` only used in `brancher` and `exp show`. Because in `exp list/push/pull/remove` we do not need to support `all-tags` and `all-branches`. Do we still need this? or we can just use `last_n_commits` plus `all-tags` and `all-branches` and `all-commits` in `brancher` and `exp show`, or maybe it is useful in some other commands (for example `gc` and `exp gc`)

Reply:
I think the real issue is that `iter_revs` is still heavy in the first place, when it probably shouldn't be.

I think what we really need is another refactor of `iter_revs` and the SCM `list_branches/list_tags/etc` functions so that they actually use iterators and not lists/dicts. (But this is not high priority right now, and still goes along with the need for a `brancher` refactor in general).

I think it is fine for `exp list/push/pull/etc` to continue using `iter_revs` rather than introducing another utility here, because eventually we should really be solving this problem in `iter_revs` rather than introducing more helper functions.

@karajan1001

Reply:
> I think it is fine for exp list/push/pull/etc to continue using iter_revs rather than introducing another utility here, because eventually we should really be solving this problem in iter_revs rather than introducing more helper functions.

Actually what I suggest is to abandon `iter_revs` and just use a lightweight `last_n_commits` function. Now argument  `all_experiments` only used in `brancher` beside this it makes the function in `dvc.scm.py` depends on something from `dvc/repo/experiments/utils/py`, while all other dependencies are in an opposite direction, I think this indicates here is a wrong place for it.

`all_branches`, `all_tags`, they only exist in `exp show` and `brancher`, and we only call them from scm
```python
        if all_branches:
            revs.extend(scm.list_branches())
            results.extend(scm.list_branches())

        if all_tags:
            revs.extend(scm.list_tags())
            results.extend(scm.list_tags())
```

The only two universal flags here are `all_commits` and `revs`. and if the command has an all_commits flag, it is better to directly collect experiments during the iteration and do not check their baselines for performance reasons. So the only logic I think should be separated here is the `rev` + `num`.

Reply:
You are probably right that `all_experiments` doesn't belong in `iter_revs`, since it's specific to brancher. But otherwise, ideally, `iter_revs` should be just as lightweight as `last_n_commits` (and it should not work by extending a list).

If you want to keep `last_n_commits` separate for now, that's fine with me, but I do think that eventually (once we revisit brancher) that functionality will end up being folded back into a refactored `iter_revs`.

For this PR though, please change `last_n_commits` to be a generator, there is no need to add the overhead of a list here